src/gtc/common.py

```diff
@@ -572,6 +572,101 @@ def apply(cls, node: Node, *, symtable: Dict[str, Any]) -> List[str]:
     return root_validator(allow_reuse=True, skip_on_failure=True)(_impl)
 
 
+class _LvalueDimsValidator(NodeVisitor):
+    def __init__(self, vertical_loop_type: Type[Node], decl_type: Type[Node]) -> None:
+        if not vertical_loop_type.__annotations__.get("loop_order") is LoopOrder:
+            raise ValueError(
+                f"Vertical loop type {vertical_loop_type} has no `loop_order` attribute"
+            )
+        if not decl_type.__annotations__.get("dimensions") is Tuple[bool, bool, bool]:
+            raise ValueError(f"Field decl type {decl_type} has no `dimensions` attribute")
+        self.vertical_loop_type = vertical_loop_type
+        self.decl_type = decl_type
+
+    def visit_Node(
+        self,
+        node: Node,
+        *,
+        symtable: Dict[str, Any],
+        loop_order: Optional[LoopOrder] = None,
+        **kwargs: Any,
+    ) -> None:
+        if isinstance(node, self.vertical_loop_type):
+            loop_order = node.loop_order
+        if isinstance(node, SymbolTableTrait):
+            symtable = {**symtable, **node.symtable_}
+        self.generic_visit(node, symtable=symtable, loop_order=loop_order, **kwargs)
+
+    def visit_AssignStmt(
+        self,
+        node: AssignStmt,
+        *,
+        loop_order: LoopOrder,
+        symtable: Dict[str, Any],
+        **kwargs: Any,
+    ) -> None:
+        if not isinstance(symtable[node.left.name], self.decl_type):
+            return None
+        allowed_flags = self._allowed_flags(loop_order)
+        flags = symtable[node.left.name].dimensions
+        if flags not in allowed_flags:
+            dims = dimension_flags_to_names(flags)
+            raise ValueError(
+                f"Not allowed to assign to {dims}-field `{node.left.name}` in {loop_order.name}."
+            )
+        return None
+
+    def _allowed_flags(self, loop_order: LoopOrder) -> List[Tuple[bool, bool, bool]]:
+        allowed_flags = [(True, True, True)]  # ijk always allowed
+        if loop_order is not LoopOrder.PARALLEL:
+            allowed_flags.append((True, True, False))  # ij only allowed in FORWARD and BACKWARD
+        return allowed_flags
+
+
+# TODO(ricoh) consider making gtir.Decl & oir.Decl common
```

Comment:
Shouldn't more of this comment fit on each line?

Reply:
I suggest making a separate PR for this after merging ;)

Reply:
done